### PR TITLE
Update the ASCII-to-EBCDIC table by using IBM-1047 codepage

### PR DESCRIPTION
**Pull request by tungld**
Signed-off-by: Tung D. Le <tung@jp.ibm.com>

### CODE REVIEW

**Comment by gongsu832** [comment]
@tungld You probably want to generate the full 256-char conversion table which should look like this:
```
{"0x00", "0x01", "0x02", "0x03", "0x37", "0x2D", "0x2E", "0x2F",
 "0x16", "0x05", "0x15", "0x0B", "0x0C", "0x0D", "0x0E", "0x0F",
 "0x10", "0x11", "0x12", "0x13", "0x3C", "0x3D", "0x32", "0x26",
 "0x18", "0x19", "0x3F", "0x27", "0x1C", "0x1D", "0x1E", "0x1F",
 "0x40", "0x5A", "0x7F", "0x7B", "0x5B", "0x6C", "0x50", "0x7D",
 "0x4D", "0x5D", "0x5C", "0x4E", "0x6B", "0x60", "0x4B", "0x61",
 "0xF0", "0xF1", "0xF2", "0xF3", "0xF4", "0xF5", "0xF6", "0xF7",
 "0xF8", "0xF9", "0x7A", "0x5E", "0x4C", "0x7E", "0x6E", "0x6F",
 "0x7C", "0xC1", "0xC2", "0xC3", "0xC4", "0xC5", "0xC6", "0xC7",
 "0xC8", "0xC9", "0xD1", "0xD2", "0xD3", "0xD4", "0xD5", "0xD6",
 "0xD7", "0xD8", "0xD9", "0xE2", "0xE3", "0xE4", "0xE5", "0xE6",
 "0xE7", "0xE8", "0xE9", "0xAD", "0xE0", "0xBD", "0x5F", "0x6D",
 "0x79", "0x81", "0x82", "0x83", "0x84", "0x85", "0x86", "0x87",
 "0x88", "0x89", "0x91", "0x92", "0x93", "0x94", "0x95", "0x96",
 "0x97", "0x98", "0x99", "0xA2", "0xA3", "0xA4", "0xA5", "0xA6",
 "0xA7", "0xA8", "0xA9", "0xC0", "0x4F", "0xD0", "0xA1", "0x07",
 "0x20", "0x21", "0x22", "0x23", "0x24", "0x25", "0x06", "0x17",
 "0x28", "0x29", "0x2A", "0x2B", "0x2C", "0x09", "0x0A", "0x1B",
 "0x30", "0x31", "0x1A", "0x33", "0x34", "0x35", "0x36", "0x08",
 "0x38", "0x39", "0x3A", "0x3B", "0x04", "0x14", "0x3E", "0xFF",
 "0x41", "0xAA", "0x4A", "0xB1", "0x9F", "0xB2", "0x6A", "0xB5",
 "0xBB", "0xB4", "0x9A", "0x8A", "0xB0", "0xCA", "0xAF", "0xBC",
 "0x90", "0x8F", "0xEA", "0xFA", "0xBE", "0xA0", "0xB6", "0xB3",
 "0x9D", "0xDA", "0x9B", "0x8B", "0xB7", "0xB8", "0xB9", "0xAB",
 "0x64", "0x65", "0x62", "0x66", "0x63", "0x67", "0x9E", "0x68",
 "0x74", "0x71", "0x72", "0x73", "0x78", "0x75", "0x76", "0x77",
 "0xAC", "0x69", "0xED", "0xEE", "0xEB", "0xEF", "0xEC", "0xBF",
 "0x80", "0xFD", "0xFE", "0xFB", "0xFC", "0xBA", "0xAE", "0x59",
 "0x44", "0x45", "0x42", "0x46", "0x43", "0x47", "0x9C", "0x48",
 "0x54", "0x51", "0x52", "0x53", "0x58", "0x55", "0x56", "0x57",
 "0x8C", "0x49", "0xCD", "0xCE", "0xCB", "0xCF", "0xCC", "0xE1",
 "0x70", "0xDD", "0xDE", "0xDB", "0xDC", "0x8D", "0x8E", "0xDF"};
```

**Comment by tungld** [comment]
> You probably want to generate the full 256-char conversion table which should look like this:

@gongsu832 thank you! how did you generate the table? I just wanted to write it in comment for future reference.

**Comment by gongsu832** [comment]
@tungld I used this many years ago in some PLX code I wrote so unfortunately I cannot remember at all how I got/created it. :-D I think __a2e_s() should be able to do it. You just need to include the rest 128 chars in the input string.

**Comment by gongsu832** [comment]
BTW, here is the reverse EBCDIC to ASCII conversion table if you ever need it.
```
{"0x00", "0x01", "0x02", "0x03", "0x9C", "0x09", "0x86", "0x7F",
 "0x97", "0x8D", "0x8E", "0x0B", "0x0C", "0x0D", "0x0E", "0x0F",
 "0x10", "0x11", "0x12", "0x13", "0x9D", "0x0A", "0x08", "0x87",
 "0x18", "0x19", "0x92", "0x8F", "0x1C", "0x1D", "0x1E", "0x1F",
 "0x80", "0x81", "0x82", "0x83", "0x84", "0x85", "0x17", "0x1B",
 "0x88", "0x89", "0x8A", "0x8B", "0x8C", "0x05", "0x06", "0x07",
 "0x90", "0x91", "0x16", "0x93", "0x94", "0x95", "0x96", "0x04",
 "0x98", "0x99", "0x9A", "0x9B", "0x14", "0x15", "0x9E", "0x1A",
 "0x20", "0xA0", "0xE2", "0xE4", "0xE0", "0xE1", "0xE3", "0xE5",
 "0xE7", "0xF1", "0xA2", "0x2E", "0x3C", "0x28", "0x2B", "0x7C",
 "0x26", "0xE9", "0xEA", "0xEB", "0xE8", "0xED", "0xEE", "0xEF",
 "0xEC", "0xDF", "0x21", "0x24", "0x2A", "0x29", "0x3B", "0x5E",
 "0x2D", "0x2F", "0xC2", "0xC4", "0xC0", "0xC1", "0xC3", "0xC5",
 "0xC7", "0xD1", "0xA6", "0x2C", "0x25", "0x5F", "0x3E", "0x3F",
 "0xF8", "0xC9", "0xCA", "0xCB", "0xC8", "0xCD", "0xCE", "0xCF",
 "0xCC", "0x60", "0x3A", "0x23", "0x40", "0x27", "0x3D", "0x22",
 "0xD8", "0x61", "0x62", "0x63", "0x64", "0x65", "0x66", "0x67",
 "0x68", "0x69", "0xAB", "0xBB", "0xF0", "0xFD", "0xFE", "0xB1",
 "0xB0", "0x6A", "0x6B", "0x6C", "0x6D", "0x6E", "0x6F", "0x70",
 "0x71", "0x72", "0xAA", "0xBA", "0xE6", "0xB8", "0xC6", "0xA4",
 "0xB5", "0x7E", "0x73", "0x74", "0x75", "0x76", "0x77", "0x78",
 "0x79", "0x7A", "0xA1", "0xBF", "0xD0", "0x5B", "0xDE", "0xAE",
 "0xAC", "0xA3", "0xA5", "0xB7", "0xA9", "0xA7", "0xB6", "0xBC",
 "0xBD", "0xBE", "0xDD", "0xA8", "0xAF", "0x5D", "0xB4", "0xD7",
 "0x7B", "0x41", "0x42", "0x43", "0x44", "0x45", "0x46", "0x47",
 "0x48", "0x49", "0xAD", "0xF4", "0xF6", "0xF2", "0xF3", "0xF5",
 "0x7D", "0x4A", "0x4B", "0x4C", "0x4D", "0x4E", "0x4F", "0x50",
 "0x51", "0x52", "0xB9", "0xFB", "0xFC", "0xF9", "0xFA", "0xFF",
 "0x5C", "0xF7", "0x53", "0x54", "0x55", "0x56", "0x57", "0x58",
 "0x59", "0x5A", "0xB2", "0xD4", "0xD6", "0xD2", "0xD3", "0xD5",
 "0x30", "0x31", "0x32", "0x33", "0x34", "0x35", "0x36", "0x37",
 "0x38", "0x39", "0xB3", "0xDB", "0xDC", "0xD9", "0xDA", "0x9F"};
```

**Comment by tungld** [comment]
Thank you, @gongsu832!

I updated the table, and add EBCDIC-to-ASCII as well in just case we need it in the future.

File: src/Conversion/KrnlToLLVM/KrnlToLLVMHelper.cpp [hunk 1]
```diff
@@ -281,5 +302,55 @@ std::string a2e_s(std::string a_s) {
...
+  while (*s) {
+    std::string hex = e2a[(int)*s];
+    unsigned char dec = (unsigned char)std::stoul(hex.c_str(), nullptr, 16);
+    r.push_back(dec);
+    s++;
+  }
+  r.push_back('\0');
+  return r;
+}
+
```
Comment (gongsu832) [inline src/Conversion/KrnlToLLVM/KrnlToLLVMHelper.cpp:354]: The function can be simplified if you just make `e2a` an `unsigned char` array.
```
  unsigned char e2a[256] = {
    0x00, 0x01, 0x02, 0x03, 0x9C, 0x09, 0x86, 0x7F,
    ...
  };
  // string constructor with copy
  std::string r(e_s);
  for(int i = 0; i < r.size(); i++)
    r[i] = e2a[r[i]];
  return r;
```

Comment (tungld) [reply src/Conversion/KrnlToLLVM/KrnlToLLVMHelper.cpp:342]: Great! It is more concise. I updated it. Thanks!

File: src/Conversion/KrnlToLLVM/KrnlToLLVMHelper.cpp [hunk 2]
```diff
@@ -253,31 +253,90 @@ FlatSymbolRefAttr getOrInsertStrncmp(OpBuilder &builder, ModuleOp module) {
...
-  while (*s) {
-    std::string hex = a2e[(int)*s];
-    unsigned char dec = (unsigned char)std::stoul(hex.c_str(), nullptr, 16);
-    r.push_back(dec);
-    s++;
-  }
-  r.push_back('\0');
+  // ASCII to EBCDIC IBM-1047.
+  // clang-format off
+  unsigned char a2e[256] = {
```
Comment (etiotto) [inline src/Conversion/KrnlToLLVM/KrnlToLLVMHelper.cpp:258]: Every time these function (and the other one `e2a_s`) is called the array needs to be initialized.  Can you make the array an global `static constexpr` array ?

**Review by caoimhinuibrian — APPROVED** [review]
LGTM, but see Ettore's comment

File: src/Conversion/KrnlToLLVM/KrnlToLLVMHelper.cpp (hunk 2 above)
Comment (tungld) [reply src/Conversion/KrnlToLLVM/KrnlToLLVMHelper.cpp]: Done. Thanks!

**Pull request merged**